pkg/backend/display/progress_test.go

```diff
@@ -176,6 +177,70 @@ func TestProgressEvents(t *testing.T) {
 	}
 }
 
+func TestCaptureProgressEvents(t *testing.T) {
+	t.Parallel()
+
+	// test events
+
+	// If we see a ResourceOperationFailed event, the update is marked as failed.
+	resourceOperationFailedEvent := engine.NewEvent(engine.ResourceOperationFailedPayload{
+		Metadata: engine.StepEventMetadata{
+			URN: "urn:pulumi:dev::eks::pulumi:pulumi:Stack::eks-dev",
+			Op:  deploy.OpUpdate,
+		},
+	})
+
+	// Combined is a common failed update which renders the diagnostic section and the Copilot summary.
+	failureEvents := []engine.Event{resourceOperationFailedEvent}
+
+	// test cases
+	testCases := []struct {
+		name string
+		// is this an update that failed
+		isFailure bool
+		// expect a summary to be printed
+		expectFailure bool
+	}{
+		// Includes a failed resource operation
+		{
+			name:          "update-failure",
+			isFailure:     true,
+			expectFailure: true,
+		},
+		// Includes a successful resource operation
+		{
+			name:          "update-success",
+			isFailure:     false,
+			expectFailure: false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Parallel()
+
+			captureRenderer := NewCaptureProgressEvents(tokens.MustParseStackName("stack"), "project", Options{})
+			eventChannel, doneChannel := make(chan engine.Event), make(chan bool)
+			go captureRenderer.ProcessEvents(eventChannel, doneChannel)
+
+			if testCase.isFailure {
+				for _, event := range failureEvents {
+					eventChannel <- event
+				}
+			}
+
+			close(eventChannel)
+			<-doneChannel
+
+			if testCase.expectFailure {
+				assert.True(t, captureRenderer.OutputIncludesFailure())
+			} else {
+				assert.False(t, captureRenderer.OutputIncludesFailure())
+			}
+		})
+	}
+}
```

Comment:
If there are only two test cases, I think this'd read easier as two separate test functions, each with a doc comment explaining what it tests. Would save all the ifs, and boilerplate etc.